devito/builtins.py

```diff
@@ -53,6 +54,131 @@ def smooth(f, g, axis=None):
+                    indgl.append(slice(0, lw, 1))
+                    indfr.append(-1)
+                    indgr.append(slice(-lw, -1, 1))
+                else:
+                    indfl.append(slice(0, -1, 1))
+                    indgl.append(slice(lw, -lw, 1))
+                    indfr.append(slice(0, -1, 1))
+                    indgr.append(slice(lw, -lw, 1))
+            if isinstance(f, np.ndarray):
+                g.data[tuple(indgl)] = f[tuple(indfl)]
+                g.data[tuple(indgr)] = f[tuple(indfr)]
+            elif isinstance(f, dv.Function):
+                g.data[tuple(indgl)] = f.data[tuple(indfl)]
+                g.data[tuple(indgr)] = f.data[tuple(indfr)]
+            else:
+                raise NotImplementedError
+        elif mode == 'reflect':
+            for d in g.grid.dimensions:
+                if d == dim:
+                    indfl.append(slice(lw-1, None, -1))
+                    indgl.append(slice(0, lw, 1))
+                    indfr.append(slice(-1, -(lw+1), -1))
+                    indgr.append(slice(-lw, None, 1))
+                else:
+                    indfl.append(slice(0, None, 1))
+                    indgl.append(slice(lw, -lw, 1))
+                    indfr.append(slice(0, None, 1))
+                    indgr.append(slice(lw, -lw, 1))
+            if isinstance(f, np.ndarray):
+                g.data[tuple(indgl)] = f[tuple(indfl)]
+                g.data[tuple(indgr)] = f[tuple(indfr)]
+            elif isinstance(f, dv.Function):
+                g.data[tuple(indgl)] = f.data[tuple(indfl)]
+                g.data[tuple(indgr)] = f.data[tuple(indfr)]
+            else:
+                raise NotImplementedError
+        else:
+            raise ValueError("Mode not available")
+
+        return
+
+    def subfloor(f, g):
+        sl = slice(lw, -lw, 1)
+        indices = []
+        slices = (slice(None, None, 1), )*len(g.grid.dimensions)
+        for _ in range(len(g.grid.dimensions)):
+            indices.append(sl)
+        if isinstance(f, np.ndarray):
+            f[slices] = np.floor(g.data[tuple(indices)])
+        elif isinstance(f, dv.Function):
+            f.data[slices] = np.floor(g.data[tuple(indices)])
+        else:
+            raise NotImplementedError
+
+    lw = int(_order*sigma + 0.5)
+
+    # Create the padded grid:
+    datadomain = DataDomain(lw)
+    # FIMXE: change to type test
+    try:
+        shape_padded = np.array(f.grid.shape) + 2*lw
+    except AttributeError:
+        shape_padded = np.array(f.shape) + 2*lw
+    grid = dv.Grid(shape=shape_padded, subdomains=datadomain)
+    dims = grid.dimensions
+
+    f_c = dv.Function(name='f_c', grid=grid, space_order=2*lw,
+                      coefficients='symbolic', dtype=np.int32)
+    f_o = dv.Function(name='f_o', grid=grid, coefficients='symbolic', dtype=np.int32)
+
+    weights = np.exp(-0.5/sigma**2*(np.linspace(-lw, lw, 2*lw+1))**2)
+    weights = weights/weights.sum()
+
+    for d in dims:
```

Comment:
Can make it in one go with a 3D kernel from outer product of 1D kernel and avoid the loop as 3D smooth is
W[i,j,k] = w[i]*w[j]*w[k] and need one 3d conv rather than multiple 1D ones with padding.

Reply:
Sure, but then the stencil is a square/cube (2d/3d). I think the symbolic coefficients stuff may need an update to get that to work, will double check. (It should of course be done that way, and will be at some point, but I don't think it'll work immediately which is why it's currently like this).

Reply:
Yes but you can just do `sum w[i]*f.subs(...)` rather than symbolic coeefs and do just all dims at once